### ISSUE TITLE
Weird loss-inconsistencies and worse accuracy compared to older versions

### ISSUE THREAD

**Issue by M-Wolff** [tensorflow/tensorflow#50244]
**System information**
- Have I written custom code: Yes
- OS Platform and Distribution: Windows 10
- Mobile device: -
- TensorFlow installed from: pip install tensorflow-gpu
- TensorFlow version: v2.4.0-49-g85c8b2a817f 2.4.1
- Python version: 3.7.9
- Bazel version: -
- GCC/Compiler version: -
- CUDA/cuDNN version: 11.3
- GPU model and memory: Nvidia GTX 1060

I am working with ResNet50 from keras and am currently unable to achieve the same results I do with TF 1.13.1 in the newer TF 2.4.1.
Firstly the weights in ResNet50 have a very different range of values compared to TF 1.13.1.
In the old version for example one layer (i.e. model.layers[-6]) with 512*2048 weights has values between -0.15 and +0.15 while TF 2.4.1 has only values between -0.05 and +0.05. I also let the he_normal initializer generate 100 random numbers in TF 1.13.1 and TF 2.4.1 each and the old version also produces a range of values that is 3 times higher than the new version's one.
**Why did that change?**

I now tried to save the weights in TF 1.13.1 and import them in TF 2.4.1 to see if this is the breaking change that makes my accuracy and training progress worse.
However when I evaluate the untrained, randomly initialized model I get weird CategoricalCrossentropy loss values from model.evaluate.

On TF 1.13.1 I executed the following code to generate the saved weights file:
```
model = keras.applications.resnet50.ResNet50(weights=None, include_top=True, classes=5)
model.compile(loss="categorical_crossentropy", metrics=['accuracy','categorical_crossentropy'])
model.save("TF-1-13-1before_training")
```

On TF 2.4.1 I loaded a custom dataset and the model:
```
model = keras.models.load_model("TF-1-13-1before_training")
print(x_train.shape)  # (10, 224, 224, 3) 10 color-images, 2 for each of the 5 classes 224x224
print(y_train.shape)  # (10, 5) 10 times a one-hot encoded vector for 5 classes

predictions = model.predict(x_train)
cce = keras.losses.CategoricalCrossentropy()

cce(y_train, predictions).eval(session=tf.compat.v1.Session())  # Yields 12.894476
model.evaluate(x=x_train, y=y_train)  # Yields [463.46978759765625, 0.2, 463.4698]
model.metrics_names  # gives ['loss', 'acc', 'categorical_crossentropy']
```
```
print(y_train)
[[1. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0.]
 [0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 0. 1. 0.]
 [0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 1.]
 [0. 0. 0. 0. 1.]]
```
```
print(predictions)
[[1. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0.]]
```

**Why does model.evaluate return different categorical_crossentropy loss values than the standalone CCE function for the same data? How can the high value of ~463 be explained in contrast to the expected value of 12.89?**
The standalone CategoricalCrossentropy function applied on predictions and target values seems correct, the return value of model.evaluate on the other hand does not.

Thank you in advance for your help

**Comment by tilakrayal** [comment]
@M-Wolff ,

In order to expedite the trouble-shooting process, could you please provide the complete code of both versions and dataset  or colab link to reproduce the issue reported here.

Thanks!

**Comment by M-Wolff** [comment]
@tilakrayal 
I created a github repo with 2 code files, one for TF 2.4.1 and one for TF 1.13.1.
Please run TF 1.13.1 first because this will save the weights of this version to the disk to be used in TF 2.4.1 code.
It also includes the output generated on my machine.
To be a bit more specific, I am planning on training on a HPC Cluster, thus the available software is somewhat limited. For running with TF 1.13.1 I have to load:
GCC/8.2.0-2.31.1
CUDA/10.1.105
OpenMPI/3.1.3
TensorFlow/1.13.1-Python-3.7.2

For TF 2.4.1 the software is:
GCC/10.2.0
CUDA/11.1.1
OpenMPI/4.0.5
TensorFlow/2.4.1


What I find odd about this issue:

- initialized weights in TF 1.13.1 are between -0.14 ... 0.14, in TF 2.4.1 just between -0.05 ... 0.05
- Metric (CCE via model.evaluate) is at 12.89 for TF 1.13.1, and at 1.61 for TF 2.4.1 with a randomly initialized Resnet50
- Predictions in TF 1.13.1 are vectors with exactly one value at 1.0 and all others at 0.0, in TF 2.4.1 the probability spreads out evenly across all classes (~0.2 for each of the 5 classes)
- Loading saved model (from TF 1.13.1) in TF 2.4.1 produces ridicoulusly high loss-values (~742) but produces the same predictions as in TF 1.13.1
- Standalone loss in TF 2.4.1 (with loaded model from Tf 1.13.1) applied on the predictions of the ResNet50 and y_train labels (categorical: i.e. [0,0,0,0,1] for class number 5) produces same loss value as model.evaluate in TF 1.13.1, but very different from model.evaluate in TF 2.4.1 for the same model.

The mentioned github repo can be found here, dataset consists of a very small subset (10 images total for 5 classes) of SwedishLeaves dataset with images in .tif format, also uploaded in my github repo:
https://github.com/M-Wolff/TF-Bug
The saved initial weights file of TF 1.13.1 is also uploaded there


Thank you for taking the time to look into this issue!

**Comment by sachinprasadhs** [comment]
Can you check with `tf.compat.v1.keras.applications.resnet50` in your TF 2 and see if the results are still different.

**Comment by M-Wolff** [comment]
I tried with `tf.compat.v1.keras.applications.resnet50.ResNet50(...)` for ResNet50 initialization and `tf.compat.v1.keras.models.load_model(...)` for loading the model of TF 1.13.1 into TF 2.4.1 and it produces the exact same results as using TF 2.4.1 without the `compat.v1` part. So no improvement there.

Code file as well as output generated on my machine are also included in the [github repo](https://github.com/M-Wolff/TF-Bug)

**Comment by M-Wolff** [comment]
Another Update:
I just tried out the same experiment without using the `ImageDataGenerator`. I simply load the images (2 per class for the training set, 50 per class for the validation set; 5 classes total) into a numpy array and start the training on ResNet50 and InceptionV3 in the old and new version.

Surprisingly TF 2.4.1 is significantly worse than the old version as this graph shows.
Top is old version, bottom is new version. Left ResNet50 and right InceptionV3.

Loss in the new version on the validation set increases, in the old version it decreases as expected. Also the range of values for the loss is very different even at the beginning of training the randomly initialized model.
![results](https://user-images.githubusercontent.com/56885160/122450664-8b63e580-cfa7-11eb-8808-e3a203367c54.png)

**Am I doing something terribly wrong in [my training script](https://github.com/M-Wolff/TF-Bug/blob/main/full_training/training_comparison.py) or is the old version of TF simply better (for small training sets)?**
When using a larger training set size (25 images for each class as training set, validation set is still 50 images per class without overlapping the train set) the differences between the version seem to matter less:
![results](https://user-images.githubusercontent.com/56885160/122453093-42616080-cfaa-11eb-8f64-9d1948fd4a10.png)

The newer version seems smoother and more stable this time with a bigger training set, however the loss values still are very different (up to 80 in new ResNet50, only ~13 in old one).

What could cause this odd behaviour?

**Comment by deeb02** [comment]
@M-Wolff , I'm reading through the thread, and I'm having trouble understanding what the bug exactly is. It's clear that there are some subtle (and some not so subtle) differences between TF1.x and TF2.x but that no surprise, and can't be totally eliminated. It's also clear that saving a model in TF1.x and loading it in TF2.x also has some subtle issues (and AFAIC is not a supported feature). Many of the examples I see here are comparing training behavior on very small datasets that can't be used to reason about modern NN optimization algorithms.

Could you please summarize the bug more precisely?

**Comment by M-Wolff** [comment]
@deeb02 Thank you for your reply. Loading old models into newer versions not being a supported feature explains the high loss values, and it seems that the weird output prediction of a randomly initialized network for ResNet50 in TF 1.13.1 (i.e. for 5 classes [0, 0, 0, 1, 0]) has been fixed in newer versions and now produces in TF 2.4.1 [0.2, 0.2, 0.2, 0.2, 0.2] as expected.

But I still wonder what change would result in TF 2.4.1 being far worse at training with few training samples than the old TF 1.13.1.
For example the swedish leaves dataset (images of swedish leaves centered on a plain white background; very high similarity within one class) performs just fine in the old version of TF with 2-4 samples per class (always above 55%) and the new version can't do much better than randomly guessing (10% accuracy for 10 classes). I do understand that this is probably an edge case and not a general bug in TensorFlow, but I would definitely say this is undesired behaviour.


So to summarize the _bug_ or weird behaviour: 
Am I correct that unless there is some severe mistake in my [code](https://github.com/M-Wolff/TF-Bug/blob/main/full_training/training_comparison.py) it might be a good approach to try the same code on the same dataset again in an older version of TF when one can't produce good results in a more recent version (especially when working on very small training sets)?

**Comment by jvishnuvardhan** [comment]
@M-Wolff This is a stale issue. 

Can you please check with recent TF versions and let us know whether this was resolved with recent TF versions? Thanks!

**Comment by M-Wolff** [comment]
@jvishnuvardhan
![new_comparison](https://user-images.githubusercontent.com/56885160/153570859-2a4509ff-5a71-4b9c-a7de-036aca3385e4.png)

I took the time to reproduce the same environment and dataset as when I created this issue
 8 months ago. By now I had already cleaned up my working directory as my project was completed months ago.

In TensorFlow 2.8.0 (latest release) the behavior still matches the one of TF 2.4.1 described in my previous posts. In TF 1.13.1 the CNN (ResNet-50 and InceptionV3) can achieve 50-70% accuracy on a test split, while in newer TF versions there is no learning progress at all, accuracy stays at 20% for 5 classes (as good as randomly guessing). Results can be compared in the picture above. The dataset consists of the Swedish-Leaf dataset (high similarity of images within a given class) with 2 images per class for training and 50 images per class for testing. 5 classes total. The code is the same one I linked earlier.

I do understand that this seems like an edge case, but in my project I trained 3770 CNN's in TF 1.13.1 and TF 2.4.1 **each** and got to the result that something weird happened somewhere between TF 1.13.1 and TF 2.4.1 (or I did something terribly wrong and nobody has noticed yet).
I analyzed the effects of some changes in the output layer of the CNN's: TF 1.13.1 and PyTorch behaved similarly and as expected, however TF 2.4.1 did not. This goes to show that something must have changed in TensorFlow that affects it's performance in a bad way as far as I can tell.

Currently it seems like if you are training on a very small dataset, and don't get good results, trying the same exact experiments in an older version of TensorFlow might be a good idea to improve the accuracies.

**Comment by jvishnuvardhan** [comment]
@M-Wolff Can you please check if you are implementing normalization twice on your inputs? I see you have a normalization in your preprocessing (normalize=True).

I will looks into this later today but if you can run normalize=False, do you get any better results? Thanks!

**Comment by M-Wolff** [comment]
@jvishnuvardhan
I used normalize=True in the image_preloader call twice, because I load training and testing data separately from different folders.
That flag should just change the intervals I get in the loaded image arrays from 0...255 to 0...1 according to the documentation of tflearn.

Interestingly, setting it to normalize=False did change my results.
The graph shows TF 2.8.0 with normalize=False, the graphs for TF 1.13.1 are the same ones I posted earlier (with normalize=True) as the goal is to improve the results in TF 2.8.0 and not make them worse in TF 1.13.1 so they match.
![new_comparison_TF280-normalize-False](https://user-images.githubusercontent.com/56885160/153697776-5c1b6990-7ee2-43ef-b920-43d601d9c8ba.png)

Focusing on the graphs for TF 2.8.0 **without** normalization (as comparison with TF 1.13.1 is now limited, since the graphs for TF 1.13.1 still show normalize=True) the following questions arise:

1. Why does the test loss go up zo 7000 or 2500 without the test accuracy changing at all (before epoch 50)?
2. test accuracy spikes up to good values in the middle, but drops just as fast again. Why?
3. The InceptionV3 CNN actually managed to go below 20% accuracy in the very end for 5 classes. It is now even worse than randomly guessing. Why is that happening?
4. The Resnet-50 CNN goes up to good accuracies as well, but then drops and ends at 25% (better as before but still not as good as TF 1.13.1). Why does it drop at all, shouldnt it stay in an upwards trend, as training tries to improve the accuracy of predictions?

**Comment by sachinprasadhs** [comment]
Could you try with different dataset and check if you notice the same behavior. 
The results in different versions are not expected even when you set the seed, there will always be minor error difference due to floating point precision error.
Additionally, you can also try with different [initializers](https://www.tensorflow.org/api_docs/python/tf/keras/initializers?version=nightly) which is best suited for your usecase.

**Comment by M-Wolff** [comment]
I am aware that I can not expect the same exact accuracy due to limited floating point precision and other factors.

Also I am not trying to improve performance in TF 2.4.1 by using an initializer that is best suited for my usecase, I am only trying to understand why switching from TF 1.13.1 to TF 2.4.1 and leaving every other parameter as it was has such a big impact on validation accuracy.
The code was producing good results in TF 1.13.1 and just by switching to TF 2.4.1, using the same exact code and datasets, the accuracy got way worse in some cases.

I trained on 5 different datasets. Other parameters that were varied in each run were the amount of classes used, size of the train-split, training from scratch vs. using pre-trained networks, different networks (ResNet-50, InceptionV3 and InceptionV3 with changes), ...
This resulted in 3770 different combinations to train neural networks on.

I did this in TF 2.4.1 and did not get the results I was expecting, so I used the same exact setup of datasets and parameters and trained using the same exact code in TF 1.13.1 and suddenly the results were better and produced trends as desired (regarding some changes in parameters).

To visualize this, here is a scatterplot showing the difference between both versions. Each dot represents the difference in validation accuracy of two **exact same** training runs in TF 2.4.1 vs. TF 1.13.1. Positive values mean that TF 2.4.1 is better, negative values that TF 1.13.1 is performing better (accuracy on test-split).
![AllDatasets](https://user-images.githubusercontent.com/56885160/162606647-8b271fa5-ffb9-4663-bb3f-ca18249d613f.png)

And yes, I double-checked, the performance in TF 2.4.1 is sometimes really 80%-90% worse (e.g. 96% accuracy in TF 1.13.1 and only 6% accuracy in TF 2.4.1 with the same setup).

General trend in that diagram is that TF 2.4.1 is in many cases performing approximately as good as TF 1.13.1 but there are cases where there is a huge difference, and those cases don't seem too rare in my analysis.
Seems to happen mostly when there is only a small amount of training images.

Restricting the scatterplot from above to only the Cifar dataset results in this scatterplot:
![Cifar](https://user-images.githubusercontent.com/56885160/162606836-7bc554b4-b742-40ac-b2c8-4e6c7aba0dfe.png)
Probably because Cifar has a lot of images, even when the training-split is limited to 10% of the images there are enough pictures present so that this weird behaviour doesn't occur. Training on larger datasets (e.g. Cifar) works well in both versions.


However the main question or issue here still persists: Training on small datasets (using very few images in the train-split) is working fine in TF 1.13.1 and in TF 2.4.1 validation accuracy decreases by a lot.

So when faced with a problem that requires training on a small dataset, it seems that one should prefer the old version above the newer TF versions. 
How can this strange behaviour be explained?

**Comment by sachinprasadhs** [comment]
Thanks for the detailed insight, could you please create this issue in keras-team/keras repo for faster resolution.
You can link this issue to the new issue as reference to all your insights.

To know more see;
https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999
Thank you!

**Comment by M-Wolff** [comment]
I opened [this issue](https://github.com/keras-team/keras/issues/16444) in keras-team/keras.
If the google-ml-butler bot doesn't decide otherwise I will comment with an update here as well, if the issue gets resolved at some point in the future.

I certainly find it a bit annoying, that the bot starts to threaten to close this issue if I don't find the time to reply within one week, especially since there was no response from your side for **7 months** at some point...

Additionally, after reading through my latest response again, I noticed that I messed up the y-axis labels in the plots. It should of course be the other way around: "TF 2.4.1 minus TF 1.13.1", so negative values actually correspond to TF 1.13.1 being better (higher accuracy). Sorry!
